README.md
# aiocogeo [![CircleCI](https://circleci.com/gh/geospatial-jeff/aiocogeo/tree/master.svg?style=svg)](https://circleci.com/gh/geospatial-jeff/aiocogeo/tree/master)[![codecov](https://codecov.io/gh/geospatial-jeff/aiocogeo/branch/master/graph/badge.svg)](https://codecov.io/gh/geospatial-jeff/aiocogeo)

## Installation
```
pip install aiocogeo

# With S3 filesystem
pip install aiocogeo[s3]
```

## Usage
COGs are opened using the `COGReader` asynchronous context manager:

```python
from aiocogeo import COGReader

async with COGReader("http://cog.tif") as cog:
    ...
```

Several filesystems are supported:
- **HTTP/HTTPS** (`http://`, `https://`)
- **S3** (`s3://`)
- **File** (`/`)

### Metadata
Generating a [rasterio-style profile](https://rasterio.readthedocs.io/en/latest/topics/profiles.html) for the COG:

```python
async with COGReader("https://async-cog-reader-test-data.s3.amazonaws.com/lzw_cog.tif") as cog:
    print(cog.profile)

>>> {'driver': 'GTiff', 'width': 10280, 'height': 12190, 'count': 3, 'dtype': 'uint8', 'transform': Affine(0.6, 0.0, 367188.0,
       0.0, -0.6, 3777102.0), 'blockxsize': 512, 'blockysize': 512, 'compress': 'lzw', 'interleave': 'pixel', 'crs': 'EPSG:26911', 'tiled': True, 'photometric': 'rgb'}
```

#### Lower Level Metadata
A COG is composed of several IFDs, each with many TIFF tags:

```python
from aiocogeo.ifd import IFD
from aiocogeo.tag import Tag

async with COGReader("https://async-cog-reader-test-data.s3.amazonaws.com/lzw_cog.tif") as cog:
    for ifd in cog:
        assert isinstance(ifd, IFD)
        for tag in ifd:
            assert isinstance(tag, Tag)
```

Each IFD contains more granular metadata about the image than what is included in the profile.  For example, finding the
tilesize for each IFD:

```python
async with COGReader("https://async-cog-reader-test-data.s3.amazonaws.com/lzw_cog.tif") as cog:
    for ifd in cog:
        print(ifd.TileWidth.value, ifd.TileHeight.value)

>>> 512 512
    128 128
    128 128
    128 128
    128 128
    128 128
```

More advanced use cases may need access to tag-level metadata:
```python
async with COGReader("https://async-cog-reader-test-data.s3.amazonaws.com/lzw_cog.tif") as cog:
    first_ifd = cog.ifds[0]
    assert first_ifd.tag_count == 24

    for tag in first_ifd:
        print(tag)

>>> Tag(code=258, name='BitsPerSample', tag_type=TagType(format='H', size=2), count=3, length=6, value=(8, 8, 8))
    Tag(code=259, name='Compression', tag_type=TagType(format='H', size=2), count=1, length=2, value=5)
    Tag(code=257, name='ImageHeight', tag_type=TagType(format='H', size=2), count=1, length=2, value=12190)
    Tag(code=256, name='ImageWidth', tag_type=TagType(format='H', size=2), count=1, length=2, value=10280)
    ...
```

### Image Data
The reader also has methods for reading internal image tiles and performing partial reads.  Currently only jpeg, lzw, deflate, packbits, and webp compressions are supported.

#### Image Tiles
Reading the top left tile of an image at native resolution:

```python
async with COGReader("https://async-cog-reader-test-data.s3.amazonaws.com/webp_cog.tif") as cog:
    x = y = z = 0
    tile = await cog.get_tile(x, y, z)
    
    ifd = cog.ifds[z]
    assert tile.shape == (ifd.bands, ifd.TileHeight.value, ifd.TileWidth.value)
```

<p align="center">
  <img width="300" height="300" src="https://async-cog-reader-test-data.s3.amazonaws.com/readme/naip_top_left_tile.jpg">
</p>


#### Partial Read
You can read a portion of the image by specifying a bounding box in the native crs of the image and an output shape:

```python
async with COGReader("https://async-cog-reader-test-data.s3.amazonaws.com/webp_cog.tif") as cog:
    assert cog.epsg == 26911
    partial_data = await cog.read(bounds=(368461,3770591,368796,3770921), shape=(512,512))
```

<p align="center">
  <img width="300" height="300" src="https://async-cog-reader-test-data.s3.amazonaws.com/readme/partial_read.jpeg">
</p>

#### Internal Masks
If the COG has an internal mask, the returned array will be a masked array:

```python
import numpy as np

async with COGReader("https://async-cog-reader-test-data.s3.amazonaws.com/naip_image_masked.tif") as cog:
    assert cog.is_masked

    tile = await cog.get_tile(0,0,0)
    assert np.ma.is_masked(tile)
```

<p align="center">
  <img src="https://async-cog-reader-test-data.s3.amazonaws.com/readme/masked_tile.jpg" width="300" />
  <img src="https://async-cog-reader-test-data.s3.amazonaws.com/readme/mask.jpg" width="300" /> 
</p>

### Configuration
Configuration options are exposed through environment variables:
- **INGESTED_BYTES_AT_OPEN** - defines the number of bytes in the first GET request at file opening (defaults to 16KB)
- **HEADER_CHUNK_SIZE** - chunk size used to read header (defaults to 16KB)
- **ENABLE_BLOCK_CACHE** - determines if image blocks are cached in memory (defaults to TRUE)
- **ENABLE_HEADER_CACHE** - determines if COG headers are cached in memory (defaults to TRUE)
- **HTTP_MERGE_CONSECUTIVE_RANGES** - determines if consecutive ranges are merged into a single request (defaults to FALSE)
- **BOUNDLESS_READ** - determines if internal tiles outside the bounds of the IFD are read (defaults to TRUE)
- **BOUNDLESS_READ_FILL_VALUE** - determines the value used to fill boundless reads (defaults to 0)
- **LOG_LEVEL** - determines the log level used by the package (defaults to ERROR)
- **VERBOSE_LOGS** - enables verbose logging, designed for use when `LOG_LEVEL=DEBUG` (defaults to FALSE)
- **AWS_REQUEST_PAYER** - set to `requester` to enable reading from S3 RequesterPays buckets.

Refer to [`aiocogeo/config.py`](https://github.com/geospatial-jeff/aiocogeo/blob/master/aiocogeo/config.py) for more details about configuration options.

## CLI
```
$ aiocogeo --help
Usage: aiocogeo [OPTIONS] COMMAND [ARGS]...

Options:
  --install-completion [bash|zsh|fish|powershell|pwsh]
                                  Install completion for the specified shell.
  --show-completion [bash|zsh|fish|powershell|pwsh]
                                  Show completion for the specified shell, to
                                  copy it or customize the installation.

  --help                          Show this message and exit.

Commands:
  create-tms  Create OGC TileMatrixSet.
  info        Read COG metadata.

```

CHANGELOG.txt
0.3.0 (2020-11-30)
------------------
- Remove unnecessary calls to the thread pool (#89)
- Add header cache (#90, #94)
   - Renames the `ENABLE_CACHE` environment variable to `ENABLE_BLOCK_CACHE`.
   - Adds the `ENABLE_HEADER_CACHE` environment variable.
- Restructure test cases (#91)
- Implement chunking and smarter tag reading to reduce number of requests sent when reading the header (#95)
   - Adds the `HEADER_CHUNK_SIZE` environment variable (defaults to 16KB).

0.2.7 (2020-11-15)
------------------
- Add `CompositeReader` and `STACReader` (#70)
- Align mask recognition with GDAL RFC 15 (#72)
- Read color interpretation (#75)
- Improve geokey parsing (#79)


0.1.7 (2020-08-15)
------------------
- Add method to read pixel values at a given point (#62)
- Add method to create downsampled version of the image (#63)
- Bugfixes (#60, #61)


0.1.6 (2020-6-24)
------------------
- Use PIL for resize instead of skimage (#57)


0.1.5 (2020-6-13)
------------------
- Add support for nodata values (#47)
- Improve filesystem exception handling (#48)
- Track header size (#49)
- Add support for boundless reads (#51)
   - Add ``BOUNDLESS_READ`` configuration option which determines if internal tiles outside bounds of the IFD are read
   - Add ``BOUNDLESS_READ_FILL_VALUE`` configuration option which determines the value used to fill boundless reads
- Remove asyncio callbacks (#52)
- Add support for packbits compressed images (#54)


0.1.4 (2020-6-07)
------------------
- Runs cpu bound code in background executor (#44)


0.1.3 (2020-06-06)
------------------
- Packages s3 filesystem as install extra (#40)
- Cache each range request instead of each internal tile (#41)


0.0.3 (2020-06-01)
------------------
- Add several configuration options as environment variables:
   - ``INGESTED_BYTES_AT_OPEN`` - determines the number of bytes in the first GET request at file opening (#27)
   - ``ENABLE_BLOCK_CACHE`` - determines if internal blocks are cached in memory (#35)
   - ``HTTP_MERGE_CONSECUTIVE_RANGES`` - determines if consecutive ranges are merged into a single request (#29, #33)
   - ``LOG_LEVEL`` - determines the log level used by the library (#36)
   - ``VERBOSE_LOGS`` - enables verbose logging (#36)
- Fix bug where internal nodata masks were not returned by a partial read (#31)


0.0.2 (2020-05-25)
------------------
- Fix packaging bug in ``setup.py``


0.0.1 (2020-05-25)
------------------
- First release!!

aiocogeo/constants.py
import enum

WEB_MERCATOR_EPSG = 3857

COMPRESSIONS = {
    1: "uncompressed",
    5: "lzw",
    6: "jpeg",
    7: "jpeg",
    8: "deflate",
    32773: "packbits",
    50001: "webp",
}

INTERLEAVE = {1: "pixel", 2: "band"}

class ColorInterp(enum.IntEnum):
    """https://github.com/mapbox/rasterio/blob/master/rasterio/enums.py#L6-L25"""
    undefined = 0
    gray = 1
    grey = 1
    palette = 2
    red = 3
    green = 4
    blue = 5
    alpha = 6
    hue = 7
    saturation = 8
    lightness = 9
    cyan = 10
    magenta = 11
    yellow = 12
    black = 13
    Y = 14
    Cb = 15
    Cr = 16

PHOTOMETRIC = {
    0: "miniswhite",
    1: "minisblack",
    2: "rgb",
    3: "palette",
    4: "mask",
    5: "cmyk",
    6: "ycbcr",
    8: "cielab",
    9: "icclab",
    10: "itulab",
}

# https://github.com/cgohlke/tifffile/blob/master/tifffile/tifffile.py#L7901-L7986
# Map SampleFormat and BitsPerSample tags to numpy dtype
SAMPLE_DTYPES = {
    # UINT
    (1, 1): "?",  # bitmap
    (1, 2): "B",
    (1, 3): "B",
    (1, 4): "B",
    (1, 5): "B",
    (1, 6): "B",
    (1, 7): "B",
    (1, 8): "B",
    (1, 9): "H",
    (1, 10): "H",
    (1, 11): "H",
    (1, 12): "H",
    (1, 13): "H",
    (1, 14): "H",
    (1, 15): "H",
    (1, 16): "H",
    (1, 17): "I",
    (1, 18): "I",
    (1, 19): "I",
    (1, 20): "I",
    (1, 21): "I",
    (1, 22): "I",
    (1, 23): "I",
    (1, 24): "I",
    (1, 25): "I",
    (1, 26): "I",
    (1, 27): "I",
    (1, 28): "I",
    (1, 29): "I",
    (1, 30): "I",
    (1, 31): "I",
    (1, 32): "I",
    (1, 64): "Q",
    # VOID : treat as UINT
    (4, 1): "?",  # bitmap
    (4, 2): "B",
    (4, 3): "B",
    (4, 4): "B",
    (4, 5): "B",
    (4, 6): "B",
    (4, 7): "B",
    (4, 8): "B",
    (4, 9): "H",
    (4, 10): "H",
    (4, 11): "H",
    (4, 12): "H",
    (4, 13): "H",
    (4, 14): "H",
    (4, 15): "H",
    (4, 16): "H",
    (4, 17): "I",
    (4, 18): "I",
    (4, 19): "I",
    (4, 20): "I",
    (4, 21): "I",
    (4, 22): "I",
    (4, 23): "I",
    (4, 24): "I",
    (4, 25): "I",
    (4, 26): "I",
    (4, 27): "I",
    (4, 28): "I",
    (4, 29): "I",
    (4, 30): "I",
    (4, 31): "I",
    (4, 32): "I",
    (4, 64): "Q",
    # INT
    (2, 8): "b",
    (2, 16): "h",
    (2, 32): "i",
    (2, 64): "q",
    # IEEEFP : 24 bit not supported by numpy
    (3, 16): "e",
    # (3, 24): '',  #
    (3, 32): "f",
    (3, 64): "d",
    # COMPLEXIEEEFP
    (6, 64): "F",
    (6, 128): "D",
    # RGB565
    (1, (5, 6, 5)): "B",
    # COMPLEXINT : not supported by numpy
}

# https://github.com/python-pillow/Pillow/blob/master/src/PIL/TiffTags.py
TIFF_TAGS = {
    254: "NewSubfileType",
    256: "ImageWidth",
    257: "ImageHeight",
    258: "BitsPerSample",
    259: "Compression",
    262: "PhotometricInterpretation",
    269: "DocumentName",
    270: "ImageDescription",
    277: "SamplesPerPixel",
    280: "MinSampleValue",
    281: "MaxSampleValue",
    282: "XResolution",
    283: "YResolution",
    284: "PlanarConfiguration",
    296: "ResolutionUnit",
    305: "Software",
    306: "DateTime",
    315: "Artist",
    316: "HostComputer",
    317: "Predictor",
    320: "ColorMap",
    322: "TileWidth",
    323: "TileHeight",
    324: "TileOffsets",
    325: "TileByteCounts",
    338: "ExtraSamples",
    339: "SampleFormat",
    347: "JPEGTables",
    33432: "Copyright",
    33550: "ModelPixelScaleTag",
    33922: "ModelTiepointTag",
    34735: "GeoKeyDirectoryTag",
    42112: "GdalMetadata",
    42113: "NoData"
}


GEO_KEYS = {
    1025: "RasterType",
    2048: "GeographicType",
    3072: "ProjectedType",
}


# https://gdal.org/drivers/raster/gtiff.html#metadata
GDAL_METADATA_TAGS = [
    "DocumentName",
    "ImageDescription",
    "Software",
    "DateTime",
    "Artist",
    "HostComputer",
    "Copyright",
    "XResolution",
    "YResolution",
    "ResolutionUnit",
    "MinSampleValue",
    "MaxSampleValue"
]


RASTER_TYPE = {
    0: "Unknown",
    1: "Area",
    2: "Point",
}


class MaskFlags(enum.IntEnum):
    """https://github.com/mapbox/rasterio/blob/master/rasterio/enums.py#L80-L84"""
    all_valid = 1
    per_dataset = 2
    alpha = 4
    nodata = 8